### PR TITLE
perf(benchmarks): measure cost of decorators, dynamic components (fixes #1479, fixes #1345)

### LINKED ISSUE #1345: Measure dynamic vs static component instantiation speed
Dynamic component instantiation must be as fast as static component instantiation in ng2. The major use-case is large table with dynamic cell content.

This issue for implementing two new benchmarks:

* A list of cell components created statically
* A list of the same cell components but created dynamically

Benchmark parameters: iterations, number of cells.

# Static benchmark

```html
<static-cell *for="#cell of row.cells"></static-cell>
```

`static-cell` is a minimal component that drops a single `div` with no bindings.

# Dynamic benchmark

```html
<dynamic-cell *for="#cell of row.cells"></dynamic-cell>
```

`dynamic-cell` is a `@DynamicComponent` that loads `static-cell` using the `DynamicComponentLoader` API.

# Success metric

The numbers between the two benchmarks are within 5% of each other.

### LINKED ISSUE #1479: Benchmark the cost of decorators
We intend to use decorators as location markers for tooltips. For that to scale to tables with thousands of cells the cost of decorator must be negligible.

### PR DESCRIPTION
**Pull request by yjbanov**

### CODE REVIEW

**Comment by matanlurey** [comment]
Curious, how does it look?

**Comment by yjbanov** [comment]
@matanlurey The numbers from my laptop running on battery are a bit noisy. Sometimes it even shows that having decorators is cheaper that not having them. Let's wait till we can run them in the lab. The dynamic components seemed to be consistently more expensive by about 50%. Again, let's see how it looks in the lab.

File: modules/benchmarks/src/costs/index.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+      lifeCycle.tick();
+    });
+  });
+}
+
+@Component({selector: 'app'})
+@View({
+  directives: [If, For, DummyComponent, DummyDecorator, DynamicDummy],
+  template: `
+    <div *if="testingPlainComponents">
```
Comment (tbosch) [inline modules/benchmarks/src/costs/index.js]: Could you use a special `TemplateLoader` that unrolls the loop without using `for`? With this, the differences should show up more cleanly.

File: modules/benchmarks/src/costs/index.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  reflector.reflectionCapabilities = new ReflectionCapabilities();
+  var size = getIntParameter('size');
+  testList = ListWrapper.createFixedSize(size);
+
+  bootstrap(AppComponent).then((ref) => {
+    var injector = ref.injector;
+    var app:AppComponent = injector.get(AppComponent);
+    var lifeCycle = injector.get(LifeCycle);
+
+    bindAction('#reset', function() {
```
Comment (tbosch) [inline modules/benchmarks/src/costs/index.js]: Instead of a `reset`, could you have 2 buttons for each case, `create` and `destroy`?
Measuring destruction is also important, and with this setup, the garbage reported by benchpress is actually meaningful, as it is only created by this one use case.

File: modules/angular2/src/dom/browser_adapter.es6 [hunk 3]
```diff
@@ -102,7 +102,7 @@ export class BrowserDomAdapter extends GenericBrowserDomAdapter {
   nodeValue(node:Node):string {
     return node.nodeValue;
   }
-  type(node:string) {
```
Comment (tbosch) [inline modules/angular2/src/dom/browser_adapter.es6]: How is this change needed for your benchmark?

File: modules/benchmarks/src/costs/index.js (hunk 1 above)
Comment (tbosch) [inline modules/benchmarks/src/costs/index.js]: Talked in person, a `for` is actually what a user will do.

File: modules/benchmarks/src/costs/index.js (hunk 2 above)
Comment (tbosch) [inline modules/benchmarks/src/costs/index.js]: Talked in person, `reset` is already destroying things.

File: modules/angular2/src/dom/browser_adapter.es6 (hunk 3 above)
Comment (yjbanov) [inline modules/angular2/src/dom/browser_adapter.es6]: Surprisingly the benchmark somehow triggered this method and failed a rtts assertion due to wrong type.

**Comment by tbosch** [comment]
@yjbanov Then you ran the benchmark in `dev` mode, i.e. with the rtts assertions, so the numbers you got were probably wrong...
